x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/EnableSpatialDistancePushdown.java

```diff
@@ -44,23 +68,70 @@ public class EnableSpatialDistancePushdown extends PhysicalOptimizerRules.Parame
+            return new FilterExec(filterExec.source(), esQueryExec, rewritten);
+        }
+        return filterExec;
+    }
+
+    private Map<NameId, StDistance> getPushableDistances(List<Alias> aliases) {
+        LinkedHashMap<NameId, StDistance> refs = new LinkedHashMap<>();
+        aliases.forEach(alias -> {
+            if (alias.child() instanceof StDistance distance && canPushSpatialFunctionToSource(distance, Map.of())) {
+                refs.put(alias.id(), distance);
+            } else if (alias.child() instanceof ReferenceAttribute ref && refs.containsKey(ref.id())) {
+                StDistance distance = refs.get(ref.id());
+                refs.put(alias.id(), distance);
+            }
+        });
+        return refs;
+    }
+
+    private PhysicalPlan rewrite(FilterExec filterExec, EvalExec evalExec, EsQueryExec esQueryExec) {
+        Map<NameId, StDistance> pushable = getPushableDistances(evalExec.fields());
+        if (pushable.isEmpty() == false) {
             // Find and rewrite any binary comparisons that involve a distance function and a literal
             var rewritten = filterExec.condition().transformDown(EsqlBinaryComparison.class, comparison -> {
                 ComparisonType comparisonType = ComparisonType.from(comparison.getFunctionType());
-                if (comparison.left() instanceof StDistance dist && comparison.right().foldable()) {
+                if (comparison.left() instanceof ReferenceAttribute r && pushable.containsKey(r.id()) && comparison.right().foldable()) {
+                    StDistance dist = pushable.get(r.id());
                     return rewriteComparison(comparison, dist, comparison.right(), comparisonType);
-                } else if (comparison.right() instanceof StDistance dist && comparison.left().foldable()) {
-                    return rewriteComparison(comparison, dist, comparison.left(), ComparisonType.invert(comparisonType));
-                }
+                } else if (comparison.right() instanceof ReferenceAttribute r
+                    && pushable.containsKey(r.id())
+                    && comparison.left().foldable()) {
+                        StDistance dist = pushable.get(r.id());
+                        return rewriteComparison(comparison, dist, comparison.left(), ComparisonType.invert(comparisonType));
+                    }
                 return comparison;
             });
             if (rewritten.equals(filterExec.condition()) == false) {
-                plan = new FilterExec(filterExec.source(), filterExec.child(), rewritten);
+                // Rewrite the order so the EVAL is last, allowing the filter to be pushed down to lucene by PushFiltersToSource
+                // TODO: Remove the distance attribute from the EVAL if it is no longer needed
+                FilterExec filter = new FilterExec(filterExec.source(), esQueryExec, rewritten);
+                return new EvalExec(evalExec.source(), filter, evalExec.fields());
```

Comment:
This looks like a bug and I got a reproducer:
```
FROM test | EVAL dist = ST_DISTANCE(x, TO_GEOPOINT(\"POINT(0 0)\")), foo = x::string | WHERE dist > 3000 AND foo == \"asdf\"

{"error":{"root_cause":[{"type":"null_pointer_exception","reason":"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null"}],"type":"null_pointer_exception","reason":"Cannot invoke \"org.elasticsearch.xpack.esql.planner.Layout$ChannelAndType.channel()\" because the return value of \"org.elasticsearch.xpack.esql.planner.Layout.get(org.elasticsearch.xpack.esql.core.expression.NameId)\" is null","suppressed":[{"type":"exception","reason":"1 further exceptions were dropped"},{"type":"task_cancelled_exception","reason":"parent task was cancelled [cancelled on failure]"}]},"status":500}
```
It's not always possible to move the whole filter down. We must carefully remove only the pushable predicates from the filter.

Looking at the effects of the optimization rule shows the problem quite nicely:
```
[2024-10-03T11:56:03,927][INFO ][o.e.x.e.o.LocalPhysicalPlanOptimizer] [runTask-0] Rule local.EnableSpatialDistancePushdown applied
ProjectExec[[x{f}#40, dist{r}#34, foo{r}#37]]                                                                                  = ProjectExec[[x{f}#40, dist{r}#34, foo{r}#37]]
\_LimitExec[1000[INTEGER]]                                                                                                     = \_LimitExec[1000[INTEGER]]
  \_FilterExec[dist{r}#34 > 3000[INTEGER] AND foo{r}#37 == [61 73 64 66][KEYWORD]]                                             !   \_EvalExec[[STDISTANCE(x{f}#40,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist, TOSTRING(x{f}#40) AS foo]]
    \_EvalExec[[STDISTANCE(x{f}#40,[1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0][GEO_POINT]) AS dist, TOSTRING(x{f}#40) AS foo]] !     \_FilterExec[SPATIALDISJOINT(x{f}#40,[1 11 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 70 a7 40][GEO_SHAPE]) AND foo{r}
      \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#41], limit[], sort[] estimatedRowSize[null]                    ! #37 == [61 73 64 66][KEYWORD]]
                                                                                                                               !       \_EsQueryExec[test], indexMode[standard], query[][_doc{f}#41], limit[], sort[] estimatedRowSize[null]

```
The results of the EVAL might still be needed for other predicates.

Reply:
I've fixed this bug in https://github.com/elastic/elasticsearch/pull/112938/commits/27a8b8cf62dad23fbd825a0bed81a7e88f1317c3